### PR TITLE
Resolve pickling and deepcopying bug with dynamically added traits

### LINKED ISSUE #2: Pickling or copying dynamic traits makes them loose their type and metadata.
pickling

```
a = HasTraits()
a.add_trait('foo', File(exists=True))
a.foo = Undefined

print a.traits()
print a.traits()['foo'].trait_type

inp = pickle.dumps(a, -1)
b = pickle.loads(inp)


print b.traits()
print b.traits()['foo'].trait_type

b.foo = 'idonotexist'
a.foo = 'idonotexist'
```

copying

```
In [68]: a = HasTraits()

In [69]: a.add_trait('foo', File(exists=True))

In [70]: a.foo = Undefined

In [71]: a.traits()['foo'].trait_type
Out[71]: <enthought.traits.trait_types.File object at 0x104809650>

In [72]: b = deepcopy(a)

In [73]: b.traits()['foo'].trait_type
Out[73]: <enthought.traits.trait_types.Python object at 0x10272b950>

In [74]: b.foo = 'idonotexist'

In [75]: a.foo = 'idonotexist'
```

### LINKED ISSUE #16: dynamic traits lose type information/validation when pickled
@dpinte: finally got back to this. the example below demonstrates our issue. we use dynamic traits a lot and then pickle and send our objects to remote execution nodes. so this bug affects us a fair bit. we currently work around this by not using the validation.
## code

``` python
def test_dynamic_traits():
    from nose.tools import assert_raises
    from traits.api import HasTraits, Int
    from cPickle import dumps, loads
    a = HasTraits()
    a.add_trait('foo', Int)
    a.foo = 1
    assign_a = lambda : setattr(a, 'foo', 'a')
    assert_raises(Exception, assign_a)
    pkld_a = dumps(a)
    unpkld_a = loads(pkld_a)
    assign_a_again = lambda : setattr(unpkld_a, 'foo', 'a')    
    assert_raises(Exception, assign_a_again)
```
## test

``` python
test_dynamic_traits()
---------------------------------------------------------------------------
AssertionError                            Traceback (most recent call last)
/<ipython-input-2-deb613781b32> in <module>()
----> 1 test_dynamic_traits()

/<ipython-input-1-3ed06dbc8b75> in test_dynamic_traits()
     11     unpkld_a = loads(pkld_a)
     12     assign_a_again = lambda : setattr(unpkld_a, 'foo', 'a')
---> 13     assert_raises(Exception, assign_a_again)
     14 

/Library/Frameworks/EPD64.framework/Versions/7.1/lib/python2.7/unittest/case.pyc in assertRaises(self, excClass, callableObj, *args, **kwargs)
    469             return context
    470         with context:
--> 471             callableObj(*args, **kwargs)
    472 
    473     def _getAssertEqualityFunc(self, first, second):

/Library/Frameworks/EPD64.framework/Versions/7.1/lib/python2.7/unittest/case.pyc in __exit__(self, exc_type, exc_value, tb)
    113                 exc_name = str(self.expected)
    114             raise self.failureException(
--> 115                 "{0} not raised".format(exc_name))
    116         if not issubclass(exc_type, self.expected):
    117             # let unexpected exceptions pass through


AssertionError: Exception not raised
```
## traits version

``` python
import traits

traits.__version__
Out[4]: '4.1.1'
```

### PR DESCRIPTION
**Pull request by cfarrow**
This is #23 with working tests.

Fixes #2
Fixes #16 
Closes #23

The fix looks good to me. Any other takers?

### CODE REVIEW

File: traits/has_traits.py [hunk 1]
```diff
@@ -1425,9 +1428,12 @@ def __setstate__ ( self, state, trait_change_notify = True ):
         else:
             # Otherwise, apply the Traits 3.0 restore logic:
             self._init_trait_listeners()
+            inst_traits = state.pop('__instance_traits__', {})
             self.trait_set( trait_change_notify = trait_change_notify, **state )
             self._post_init_trait_listeners()
             self.traits_init()
+            for attr, trait in inst_traits.iteritems():
```
Comment (mdickinson) [inline traits/has_traits.py]: @cfarrow I notice that here we're calling `add_trait` _after_ doing `traits_init()` and `_post_init_trait_listeners()`, but in the `clone_traits` block below, we're calling `add_trait` _before_ those calls. Do you know whether there's a reason for the difference?

File: traits/tests/test_regression.py [hunk 2]
```diff
@@ -189,6 +190,22 @@ def test_delegation_refleak(self):
         # All the counts should be the same.
         self.assertEqual(counts[warmup:-1], counts[warmup+1:])
 
+    def test_hastraits_pickle_deepcopy(self):
```
Comment (mdickinson) [inline traits/tests/test_regression.py]: It looks as though this is a single test that tests both the pickling and the deepcopy behaviour in one. I think it should be split into two separate tests.

File: traits/has_traits.py (hunk 1 above)
Comment (cfarrow) [reply traits/has_traits.py]: I don't know the reason. I'll look into the potential consequences. At the very least I'll make it consistent.

File: traits/tests/test_regression.py (hunk 2 above)
Comment (cfarrow) [reply traits/tests/test_regression.py]: I agree.

File: traits/has_traits.py (hunk 1 above)
Comment (cfarrow) [reply traits/has_traits.py]: According to the [doc string](https://github.com/enthought/traits/blob/master/traits/ctraits.c#L1440), `traits_init` should be called at the end of `__setstate__`.  In practice, I don't think it matters. It is a no-op by default. The only place I have found it defined in the traits source is in [HasUniqueStrings](https://github.com/enthought/traits/blob/master/traits/ustr_trait.py#L141). 

In any case, I moved it after all the `add_traits` to be consistent with the existing documentation and the behavior in `clone_traits`.

**Comment by cfarrow** [comment]
Thanks for the feedback. I've addressed the comments.

**Comment by cfarrow** [comment]
Ah, python 3 test failures! I'm investigating.

**Comment by cfarrow** [comment]
Everything is working now.

File: traits/has_traits.py (hunk 1 above)
Comment (mdickinson) [reply traits/has_traits.py]: Thanks, Chris. I also moved the `_post_init_trait_listeners` call, so that it's consistent with `clone_traits`.

**Comment by mdickinson** [comment]
Waiting for CI, then merging.

**Comment by mdickinson** [comment]
Random failure at egg-fetching stage with Appveyor; restarting the job ...

**Comment by mdickinson** [comment]
Merged. Thanks, @cfarrow!

**Pull request merged**